### PR TITLE
Bump level

### PR DESCRIPTION
**Pull request by mjbain151**
I added the ability to select the SemVer release level by adding the --level option.  One of my goals during implementation was to perverse the existing interface so that the code is backwards compatible with existing code.  The level option defaults to patch release so the program should behave exactly as it did before with no level option.  Please review and provide feedback.  Thanks

### CODE REVIEW

**Comment by logicminds** [comment]
Please add change to the Changelog.md file

**Comment by mjbain151** [comment]
Updated the changelog.

File: lib/release_manager/release.rb [hunk 1]
```diff
@@ -6,25 +6,30 @@ class Release
...
   # @returns [String] the version found in the metadata file
   def version
      dry_run? ? next_version : puppet_module.version
   end
 
+  # @returns [String] the release level.  Defaults to patch
+  def level
+     options[:level] || 'patch'
+  end
+
```
Comment (logicminds) [inline lib/release_manager/release.rb]: I would put this in the cli code

`options[:level] ||= 'patch'`

File: lib/release_manager/cli/release_mod_cli.rb [hunk 2]
```diff
@@ -38,6 +38,9 @@ def self.run
         opts.on('-a', '--auto', 'Run this script without interaction') do |c|
           options[:auto] = c
         end
+        opts.on('-l', '--level', 'Semantic versioning level to bump (major.minor.patch), defaults to patch') do |c|
+          options[:level] = c
+        end
         opts.on('-m', '--module-path [MODULEPATH]', "The path to the module, defaults to #{Dir.getwd}") do |c|
           options[:path] = c
         end
```
Comment (logicminds) [inline lib/release_manager/cli/release_mod_cli.rb]: Put this here: `options[:level] ||= 'patch'` after the opts.on

**Comment by mjbain151** [comment]
Agreed.  I will make the changes.

File: lib/release_manager/cli/release_mod_cli.rb (hunk 2 above)
Comment (logicminds) [reply lib/release_manager/cli/release_mod_cli.rb]: In order to store the option you have use the following syntax.  `opts.on('-l', '--level [LEVEL]', 'blah blah'`

Also in your description please separate the levels with commas `( major,minor,patch)`

Additionally, you should also supply some validation  %w(major minor patch).include?(c)

File: lib/release_manager/release.rb [hunk 3]
```diff
@@ -41,7 +46,9 @@ def bump
       return
     end
     raise TagExists.new("Tag v#{version} already exists") if puppet_module.tag_exists?("v#{next_version}", options[:remote])
-    version = puppet_module.bump_patch_version unless options[:bump]
+    if puppet_module.respond_to?("bump_#{level}_version")
```
Comment (logicminds) [inline lib/release_manager/release.rb]: in order to check if a method exists you must convert it to a symbol.  This is the case for respond_to? and send()

`"bump_#{options[:level]}_version".to_sym`

Comment (mjbain151) [reply lib/release_manager/release.rb]: Not sure I understand.  If level is a property containing the string value  "patch".  Then I would expect "bump_#{level}_version to return the string 'bump_patch_version'

**Comment by logicminds** [comment]
https://github.com/nwops/release_manager/tree/instructions#develpment-environment-setup

**Comment by mjbain151** [comment]
I don't have docker installed.  So I will get that setup,  I don't know how
long that will take but will respond when I am done.

**Comment by mjbain151** [comment]
So I got docker installed,  but executing docker-compose up  fails.

[b1mjbcd@b1lxdev02 release_manager]$ sudo docker version
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:06 2017
 OS/Arch:      linux/amd64
Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 20:00:25 2017
 OS/Arch:      linux/amd64
 Experimental: false
[b1mjbcd@b1lxdev02 release_manager]$ docker-compose up
-bash: docker-compose: command not found

**Comment by logicminds** [comment]
Did you download docker for windows?

**Comment by mjbain151** [comment]
No my development server is a Linux server running Centos.7

**Comment by logicminds** [comment]
You will need to install docker-compose as well.

**Comment by mjbain151** [comment]
I installed docker-compose as you suggested.  But is running into an issue
that others has encountered,  see below.  I status the docker deamon and it
is running.  So I am going to try different things to see if I can get to
work.

[b1mjbcd@b1lxdev02 release_manager]$ docker-compose up
ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket -
is it running?
If it's at a non-standard location, specify the URL with the DOCKER_HOST
environment variable.

File: lib/release_manager/release.rb (hunk 3 above)
Comment (GrammatonKlaric) [reply lib/release_manager/release.rb]: @mjbain151 See this doc [respond_to](https://ruby-doc.org/core-2.4.2/Object.html#method-i-respond_to-3F). Also check out [this nice article on ruby symbols](http://www.troubleshooters.com/codecorn/ruby/symbols.htm#_How_are_symbols_implemented)

**Comment by mjbain151** [comment]
So I resolve the issue with docker-compose not connecting to the docker
daemon.  That got me further along, but now I have a new issue to resolve.

[b1mjbcd@b1lxdev02 release_manager]$ docker-compose up
Creating network "releasemanager_default" with the default driver
Pulling runner (
registry.gitlab.com/gitlab-org/gitlab-ci-multi-runner:ubuntu-v9.3.0)...
ERROR: Get
https://registry.gitlab.com/v2/gitlab-org/gitlab-ci-multi-runner/manifests/ubuntu-v9.3.0:
denied: access forbidden

**Comment by logicminds** [comment]
I failed to mention that you needed some updates.  I pushed my changes to master.  Please rebase and try docker-compose up again.  There were some changes that you did not have.

**Comment by mjbain151** [comment]
So I did what you said and WOW! the magic started.  But alas it came to
halt when runner_1 started to throw warnings, see attached.

[image: Inline image 2]

**Comment by mjbain151** [comment]
A short while later the errors went away and I was able to log into
GitLab-CE in the docker container.  So I have some exploring to do...
Ultimately I will need to get back to testing Release-manager.

[image: Inline image 1]

**Comment by mjbain151** [comment]
Added input validation for the --level parameter and squashed my commits.

**Comment by mjbain151** [comment]
Made some progress with docker.  But  I am getting there error below.
appuser@de96629fe060:/app$ bash app_startup_script.sh
Using rake 12.1.0
Using ast 2.3.0
Using bundler 1.16.0.pre.2
Using coderay 1.1.2
Using diff-lcs 1.3
Using multi_xml 0.6.0
Using httparty 0.15.6
Errno::EACCES: Permission denied @ rb_sysopen -
/usr/local/bundle/bin/httparty
An error occurred while installing httparty (0.15.6), and Bundler cannot
continue.
Make sure that `gem install httparty -v '0.15.6'` succeeds before bundling.
In Gemfile:
  release_manager was resolved to 0.5.3, which depends on
    gitlab was resolved to 4.2.0, which depends on
      httparty
Identity added: /home/appuser/.ssh/id_rsa (rsa w/o comment)

**Comment by logicminds** [comment]
Yea I was getting that error too.  I thought I resolved it.  Also your commits message needs to be summarized.  Please follow this as a guide: https://docs.puppet.com/forge/contributing.html#making-changes

**Comment by mjbain151** [comment]
Lots of interesting tips in the guide.  But I don't think I found anything
on how to summarize your commit messages.  In fact I saw interesting
articles on "Commit early and often" and "On the sausage making",   Were
you referring to squashing my commits?

**Comment by logicminds** [comment]
Yes please summary your 15 or so commit messages into a few sentences.  What does this PR provide and why it didn't work before.

**Comment by mjbain151** [comment]
Amended the commits into a single commit summary.  Thanks.

**Pull request merged**